tsdb/index/postings.go

```diff
@@ -397,6 +397,35 @@ func (p *MemPostings) addFor(id storage.SeriesRef, l labels.Label) {
 	}
 }
 
+func (p *MemPostings) PostingsForMatcher(ctx context.Context, m *labels.Matcher) Postings {
+	p.mtx.RLock()
+
+	e := p.m[m.Name]
+	if len(e) == 0 {
+		p.mtx.RUnlock()
+		return EmptyPostings()
+	}
+
+	count := 0
+	for v, srs := range e {
+		if m.Matches(v) && len(srs) > 0 {
+			count++
+		}
+	}
+
+	its := make([]Postings, 0, count)
+	for val, srs := range e {
+		if !m.Matches(val) || len(srs) == 0 {
+			continue
+		}
+
+		its = append(its, NewListPostings(srs))
+	}
```

Comment:
Style nit: both loops do the same, but use a different scheme and variable names, I'd unify that to reduce congnitive load. i.e.:

```suggestion
	for v, srs := range e {
		if m.Matches(v) && len(srs) > 0 {
			count++
		}
	}

	its := make([]Postings, 0, count)
	for v, srs := range e {
		if m.Matches(v) && len(srs) > 0 {
			its = append(its, NewListPostings(srs))
		}
	}
```